src/python/pants/backend/build_files/fmt/buildifier/rules.py

```diff
@@ -32,7 +32,30 @@ async def buildfier_fmt(
                 argv=[buildifier_tool.exe, "-type=build", *request.files],
                 input_digest=input_digest,
                 output_files=request.files,
-                description=f"Run buildifier on {pluralize(len(request.files), 'file')}.",
+                description=f"Run {Buildifier.options_scope} on {pluralize(len(request.files), 'file')}.",
+                level=LogLevel.DEBUG,
+            )
+        ),
+    )
+    return await FmtResult.create(request, result)
+
+
+# Note - this function is kept separate because it is invoked from update_build_files.py, but
+# not as a rule.
```

Comment:
Could we either invoke it as a rule, or have `buildfier_fmt` call this function to avoid the duplication?

Reply:
Yeah, I think having `buildifier_fmt` call this function is probably best - this is how the `ruff` backend works too:

https://github.com/pantsbuild/pants/blob/main/src/python/pants/backend/build_files/fmt/ruff/register.py#L19-L22

https://github.com/pantsbuild/pants/blob/main/src/python/pants/backend/python/lint/ruff/format/rules.py#L69-L71

Reply:
Oh of course, this has been now refactored.

Reply:
And now I recalled why it was duplicate - there seems to be an issue to have a call to a function which invokes other rules:

```
E       native_engine.IntrinsicError: Call(pants.core.util_rules.external_tool.download_external_tool, DownloadedExternalTool) was not detected in your @rule body at rule compile time.
```

This must a Pants architecture constraint. I'll poke around a bit more.

Leaving the duplicate code as is for now.

Reply:
Gotcha - I think this may be an issue with the new call-by-name syntax rather than the prior version of the engine. Good to know on a go-forward basis, and I think we'll need to take this into account for the other build files formatters.

Reply:
Huh, okay.

Happy to land this as is to keep moving forward, but it'd be good to have a follow-up that resolves the duplication... potentially even just deleting the `@rule` entirely, since it's apparently not used for anything (as evidenced by the backend not being wired up/not doing anything)!

That said, I believe it is be normal and supported for `@rule`s to call `async def` functions, so if call-by-name doesn't support it, we should know about it and probably fix it, e.g. random example `await _get_relevant_source_files`:

https://github.com/pantsbuild/pants/blob/459d39bb39f1f6bef605fd3b4fa129c5d721ea6f/src/python/pants/backend/javascript/install_node_package.py#L65-L88

As part of the follow-up can you investigate more deeply and maybe file a bug?

Reply:
Thanks, Huon. Definitely, let me either do a follow-up PR with either a clean up or refactoring or filing an issue.